README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# emstreeR

<!-- # emstreeR <img src="man/figures/logo.png" align="right" /> -->

<!-- [![Downloads](http://cranlogs.r-pkg.org/badges/emstreeR?color=brightgreen)](http://www.r-pkg.org/pkg/emstreeR) -->

<!-- one space after links to display badges side by side -->

[![Travis-CI Build
Status](https://travis-ci.org/allanvc/emstreeR.svg?branch=master)](https://travis-ci.org/allanvc/emstreeR)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/emstreeR)](https://cran.r-project.org/package=emstreeR)
[![Downloads from the RStudio CRAN
mirror](https://cranlogs.r-pkg.org/badges/grand-total/emstreeR)](https://cran.r-project.org/package=emstreeR)
[![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg?style=flat-square)](https://opensource.org/licenses/BSD-3-Clause)

## Overview

**emstreeR** is a package for fast and easily computing Euclidean
Minimum Spanning Trees (EMST). It heavily relies on ‘RcppMLPACK’ and
‘Rcpp’, working as a wrapper to the fast EMST Dual-Tree Boruvka
algorithm (March, Ram, Gray, 2010) implemented in ‘mlpack’ - the C++
Machine Learning library (Curtin, 2013). Thus, do not have to deal with
the R-‘Rcpp’-C++ integration. The package also provides functions and an
S3 method for readily plotting Minimum Spanning Trees (MST) using either
‘base’ R, ‘scatterplot3d’ or ‘ggplot2’ style.

  - `computeMST()` computes an Euclidean Minimum Spanning Tree for the
    input data.
  - `plot.MST()` an S3 method of the generic function plot() for
    plotting a 2D MST.
  - `plotMST3D()` plots a 3D MST using the ‘scatterplot3d’ style.
  - `stat_MST()` a ‘ggplot2’ Stat extension for plotting a 2D MST.

## Installation

``` r
# CRAN version
install.packages("emstreeR")

# Dev version
if (!require('devtools')) install.packages('devtools')
devtools::install_github("allanvc/emstreeR")
```

## Basic Usage

``` r
## artificial data:
set.seed(1984)
n <- 7
c1 <- data.frame(x = rnorm(n, -0.2, sd = 0.2), y = rnorm(n, -2, sd = 0.2))
c2 <- data.frame(x = rnorm(n, -1.1, sd = 0.15), y = rnorm(n, -2, sd = 0.3)) 
d <- rbind(c1, c2)
d <- as.data.frame(d)

## MST:
library(emstreeR)
out <- ComputeMST(d)
#> 8 edges found so far.
#> 182 cumulative base cases.
#> 0 cumulative node combinations scored.
#> 12 edges found so far.
#> 344 cumulative base cases.
#> 0 cumulative node combinations scored.
#> 13 edges found so far.
#> 442 cumulative base cases.
#> 0 cumulative node combinations scored.
#> Total spanning tree length: 8.81192
out
#>               x         y from to  distance
#> 1  -0.118159357 -2.166545   11 13 0.2039000
#> 2  -0.264604994 -2.105242    6  7 0.3429154
#> 3  -0.072829535 -1.716803   10 14 0.3540068
#> 4  -0.569225757 -1.943598    8 12 0.3541008
#> 5  -0.009270527 -1.942413    1  2 0.4825166
#> 6   0.037697969 -1.832590    3  7 0.5072094
#> 7  -0.091509110 -1.795213   12 13 0.6017045
#> 8  -1.097338236 -1.871078    5  6 0.7142135
#> 9  -0.841400898 -2.194585    8 14 0.8351710
#> 10 -1.081888729 -1.728982    4 12 0.9724635
#> 11 -1.366334073 -2.003965    4  5 1.0563892
#> 12 -1.081078171 -1.925745    2  5 1.1558933
#> 13 -1.357063682 -1.972485    2  9 1.2314331
#> 14 -0.913706515 -1.753315    1  1 0.0000000
```

## Plotting

### 2D Plots

``` r
## artifical data for 2D plots:
set.seed(1984)
n <- 15
c1 <- data.frame(x = rnorm(n, -0.2, sd = 0.2), y = rnorm(n, -2, sd = 0.2))
c2 <- data.frame(x = rnorm(n, -1.1, sd = 0.15), y = rnorm(n, -2, sd = 0.3)) 
d <- rbind(c1, c2)
d <- as.data.frame(d)
  
## MST:
library(emstreeR)
out <- ComputeMST(d, verbose = FALSE)
```

``` r
## simple 2D plot:
plot(out, col.pts = "red", col.segts = "blue")
```

<img src="man/figures/README-plot-1.png" width="650" height="500">

``` r
## 2D plot with ggplot2:
library(ggplot2)
ggplot(data = out, aes(x = x, y = y, from = from, to = to))+ 
  geom_point()+ 
  stat_MST(colour="red")
```

<img src="man/figures/README-statMST-1.png" width="600" height="400">

``` r
## 2D curved edges plot with ggplot2:
library(ggplot2)
ggplot(data = out, aes(x = x, y = y, from = from, to = to))+ 
  geom_point()+ 
  stat_MST(geom="curve")
```

<img src="man/figures/README-statMST_curved-1.png" width="600" height="400">

### 3D Plot

``` r
## artificial data for 3D plots:
n = 99
set.seed(1984)
d1<-matrix(rnorm(n, mean = -2, sd = .5), n/3, 3) # 3d
d2<-matrix(rnorm(n, mean = 0, sd = .3), n/3, 3)
d3<-matrix(rnorm(n, mean = 3, sd = .4), n/3, 3)
d<-rbind(d1,d2,d3) # just to show a matrix input
  
## MST:
library(emstreeR)
out <- ComputeMST(d, verbose = FALSE)
```

``` r
## simple 3D plot:
plotMST3D(out, xlab = "xaxis", col.pts = "orange", col.segts = "red", main = "Just a MST 3D plot")
```

<img src="man/figures/README-plotMST3D-1.png" width="600" height="400">

## Extras

``` r
## plotting MST on maps:
# honeymoon cruise example

library(ggmap)
library(dplyr)  
    
## define ports:
df.port_locations <- data.frame(location = c("Civitavecchia, Italy", 
                                             "Genova, Italy",
                                             "Marseille, France",
                                             "Barcelona, Spain",
                                             "Tunis, Tunisia",
                                             "Palermo, Italy"), 
                                stringsAsFactors = FALSE)
    
## get latitude and longitude:
geo.port_locations <- geocode(df.port_locations$location, source = "dsk")
    
## combine data:
df.port_locations <- cbind(df.port_locations, geo.port_locations)
    
## MST:
library(emstreeR)
out <- ComputeMST(df.port_locations[,2:3], verbose = FALSE)
```

``` r
## Plot:
map_grid <- c(left = -8, bottom = 32, right = 20, top = 47)
    
get_stamenmap(map_grid, zoom = 5) %>% ggmap()+
  stat_MST(data = out,
           aes(x = lon, y = lat, from=from, to=to), 
           colour="red", linetype = 2)+
  geom_point(data = out, aes(x = lon, y = lat), size=3)
```

<img src="man/figures/README-statMST_ggmaps-1.png" width="600" height="400">

## License

This package is licensed under the terms of the BSD 3-clause License.

## References

March, W. B., and Ram, P., and Gray, A. G. (2010). *Fast euclidian
minimum spanning tree: algorithm analysis, and applications*. 16th ACM
SIGKDD International Conference on Knowledge Discovery and Data mining,
July 25-28 2010. Washington, DC, USA. <doi:10.1145/1835804.1835882>.

Curtin, R. R. et al. (2013). Mlpack: A scalable C++ machine learning
library. *Journal of Machine Learning Research*, v. 14, 2013.


NEWS.md
# emstreeR 2.1.2 (2019-03-13 - third CRAN patch)

## New data, functions, and features

* updated DESCRIPTION file.
* fixed citations in DESCRIPTION, README and emstreeR-package.Rd
* updated documentation in the ComputeMST.R file 

---

# emstreeR 2.1.1 (2019-01-19 - second CRAN patch)

## Internal

* Fixed Imports in DESCRIPTION file, removing the mention to RcppMLPACK in order
to eliminate NOTES when building on Fedora and Solaris

## New data, functions, and features

* updated DESCRIPTION file.
* updated 'emstreeR-package' file.
* updated 'README.md' file.

---

# emstreeR 2.1.0 (2018-12-06 - first CRAN patch)

## Internal

* Fixed compilation in OS X and Solaris:
  - copied 'Makervars' and 'Makevars.win' from RcppMLPACK with minor modifications
  - added //Rcpp[[depends(RcppMLPACK)]] to mlpack_mst.cpp file

## New data, functions, and features

* fixed typos in the DESCRIPTION file.

---

# emstreeR 2.0.0 (2018-11-28 - pre CRAN release)

## Internal

* Fix [-Wlang-lang] windows compilation WARNINGs when running devtools::win_check() 
  by adding `CXX_STD = CXX_11` line to the Makevars.win file. Reference: "Writing
  R Extensions" manual, section 1.2.4 - "Using C++11 code".

## New data, functions, and features

* Substituted the wrapper function `plotMST()` for the S3 method `plot.MST()`

* Added 'dontrun' example 'honeymoon cruise' to the package README file 

* Removed broken Japan 'dontrun' example from the package help page 
  'emstreeR-package.Rd'because of the token problem with GoogleMaps API.

* fixed errors in the DESCRIPTION file as suggested by CRAN

R/plot.MST.R
#' @title Plot method for 'MST' objects 
#' 
#' @description Plots a 2D Minimum Spanning Tree (MST) by producing a scatter 
#' plot with segments using the generic function \code{\link{plot}}.
#'     
#' @param x a \code{MST} class object returned by the 
#'     \code{\link[emstreeR]{ComputeMST}} function.
#' @param V1 the numeric position or the name of the column to be used as the x 
#'     coordinates of the points in the plot.
#' @param V2 the numeric position or the name of the column to be used as the y 
#' coordinates of the points in the plot.
#' @param col.pts color of the points (vertices/nodes) in the plot.
#' @param col.segts color of the segments (edges) in the plot.
#' @param lty line type. An integer or name: \code{0 = "blank"}, 
#'     \code{1 = "solid"}, \code{2 = "dashed"}, \code{3 = "dotted"}, 
#'     \code{4 = "dotdash"}, \code{5 = "longdash"}, \code{6 = "twodash"}. The 
#'     default for \code{'MST'} objects is \code{"dotted"}. 
#'
#' @param ... further graphical parameters.
#' 
#' @examples
#' 
#' ## 2D artifical data
#' set.seed(1984)
#' n <- 15
#' c1 <- data.frame(x = rnorm(n,-0.2, sd=0.2), y = rnorm(n,-2,sd=0.2))
#' c2 <- data.frame(x = rnorm(n,-1.1, sd=0.15), y = rnorm(n,-2,sd=0.3)) 
#' c3 <- c(0.55,-2.4)
#' d <- rbind(c1, c2, c3)
#' d <- as.data.frame(d)
#' 
#' ## MST:
#' out <- ComputeMST(d)
#' out
#' 
#' ## 2D plot:
#' plot(out)
#' 
#' # using different parameters
#' plot(out, col.pts = "blue", col.segts = "red", lty = 2)
#' 
plot.MST <- function(x, ..., V1 = 1, V2 = 2, col.pts = "black", 
                     col.segts = "black", lty=3){
  # user can choose which columns to plot by assigning other values to V1 and V2
  tree <- x
  
  from <- tree$from
  to <- tree$to
  
  V1 = V1
  V2 = V2
  
  plot(x = tree[, V1], y = tree[, V2], col = col.pts, ...)
  segments(tree[from, V1], tree[from, V2], tree[to, V1], tree[to, V2], 
           col = col.segts, lty = lty, ...)
}


R/plotMST3D.R
#' @title 3D Minimum Spanning Tree Plot
#' 
#' @description Plots a 3D MST by producing a point cloud with segments as a 
#'     'scatterplot3d' graphic.
#' 
#' @param tree a \code{MST} class object returned by the \code{ComputeMST()} 
#'     function.
#' @param x the numeric position or the name of the column to be used as the x 
#'     coordinates of points in the plot.
#' @param y the numeric position or the name of the column to be used as the y 
#' coordinates of points in the plot.
#' @param z the numeric position or the name of the column to be used as the z 
#' coordinates of points in the plot.
#' @param col.pts color of points (vertices/nodes) in the plot.
#' @param col.segts color of segments (edges) in the plot.
#' @param angle angle between x and y axis (Attention: result depends on 
#'     scaling).
#' @param ... further graphical parameters.
#' @return NULL
#' 
#' @examples
#' 
#' ## 3D artificial data:
#' n1 = 12
#' n2 = 22
#' n3 = 7
#' n = n1+n2+n3
#' set.seed(1984)
#' 
#' mean_vector <- sample(seq(1, 10, by=2), 3)
#' sd_vector <- sample(seq(0.01, 0.8, by=0.01), 3)
#' c1 <- matrix(rnorm(n1*3,mean=mean_vector[1],sd=.3), n1, 3)
#' c2 <- matrix(rnorm(n2*3,mean=mean_vector[2],sd=.5), n2, 3)
#' c3 <- matrix(rnorm(n3*3,mean=mean_vector[3],sd=1), n3, 3)
#' d<-rbind(c1, c2, c3)
#' 
#' ## MST:
#' out <- ComputeMST(d)
#' 
#' ## 3D PLOT:
#' plotMST3D(out)
#' 
#' @export
#' 
plotMST3D <- function(tree, x=1, y=2, z=3, 
                      col.pts = "black", col.segts = "black", angle=40, ...) {
  
  x = x
  y = y
  z = z
  
  g <- scatterplot3d::scatterplot3d(tree[, x], tree[, y], tree[, z], 
                                    color=col.pts, angle=angle, ...)
  
  from <- g$xyz.convert(tree[tree$from, x] , tree[tree$from, y], 
                        tree[tree$from, z])
  to <- g$xyz.convert(tree[tree$to, x] , tree[tree$to, y], 
                      tree[tree$to, z])
  
  segments(from$x, from$y, to$x, to$y, col=col.segts, ...)
}

R/ComputeMST.R
#' @title Euclidean Minimum Spanning Tree
#'
#' @description Computes an Euclidean Minimum Spanning Tree (EMST) from the 
#'     data. \code{ComputeMST} is a wrapper for the homonym function in
#'     the 'mlpack' library.
#'
#' @param x a \code{numeric matrix} or \code{data.frame}.
#' @param verbose If \code{TRUE}, mutes the output from the C++
#'     code.
#'
#' @details Before the computation, ComputeMST runs some checks and 
#'     transformations (if needed) on the provided data using the 
#'     \link{data_check} function. After the computation, it returns the 
#'     'cleaned' data plus 3 columns: \code{from}, \code{to}, and 
#'     \code{distance}. Those columns show each pair of start and end points, 
#'     and the distance between them, forming the Minimum Spanning Tree (MST).
#'
#' @return an object of class \code{MST} and \code{data.frame}.
#'
#' @note It is worth noting that the afore mentioned  columns (\code{from}, 
#'     \code{to}, and \code{distance}) have no relationship with their 
#'     respective row in the output \code{MST/data.frame} object. The authors 
#'     chose the \code{data.frame} format for the output rather than a 
#'     \code{list} because it is more suitable for plotting the MST with the 
#'     new 'ggplot2' Stat (\code{\link[emstreeR]{stat_MST}}) provided with this 
#'     package. The last row of the output at these three columns will always 
#'     be the same: \code{1 1 0.0000000}. This is because we always have n-1 
#'     edges for n points. Hence, this is done to 'complete' the data.frame 
#'     that is returned.
#'
#' @examples
#' 
#' ## artifical data
#' set.seed(1984)
#' n <- 15
#' c1 <- data.frame(x = rnorm(n, -0.2, sd = 0.2), y = rnorm(n, -2, sd = 0.2))
#' c2 <- data.frame(x = rnorm(n, -1.1, sd = 0.15), y = rnorm(n, -2, sd = 0.3)) 
#' d <- rbind(c1, c2)
#' d <- as.data.frame(d)
#' 
#' ## MST:
#' out <- ComputeMST(d)
#' out
#' 
#' @export
#' 
ComputeMST <- function(x, verbose=TRUE) {
  # returns a mst/df object
  
  data_aux <- data_check(x)
  if(!isTRUE(verbose)) {
    BBmisc::suppressAll(
      from_to_matrix <- mlpack_mst(scale(data_aux)) # always scaled  
    )} else {
      from_to_matrix <- mlpack_mst(scale(data_aux)) # always scaled  
  } # as suggested at: https://github.com/rcppmlpack/RcppMLPACK1/issues/7
  
  
  from <- c(as.integer(from_to_matrix[1,]), 
            1 ) # repeating the last link 
  #.. for plots with ggplot2
  to <- c(as.integer(from_to_matrix[2,]), 
          1 ) # it should not be 
  #.. "from" 1 "to" 1 in order to get stat_MST working
  distance <- c(from_to_matrix[3,], 0) # adding a zero to the last row's 
  #.. distance to not intefere in the total cost of the MST
  
  colnames(data_aux) <- colnames(x)
  # we can't recover the rownames because sometimes data_check() may remove 
  #.. lines with NA in the data
  
  x <- data.frame(data_aux, from, to, distance) # from and to are not linked 
  #..to the rows
  # we just cbind from and to in order to have a data.frame and not a list
  
  rm(data_aux) # cleaning
  class(x) <- c("MST", "data.frame") # because we wanted a new plot.method for 
  #..objects of class "MST".
  
  return(x)
}


src/mlpack_mst.cpp
#include <RcppMLPACK.h>				// MLPACK, Rcpp and RcppArmadillo


using namespace mlpack::emst;
using namespace std;


//' @title \code{mlpack}'s Euclidean Minimum Spanning Tree
//'
//' @description Fast computes an EMST using the \code{mlpack C++} library.
//'
//' @param data a \code{matrix}.
//' @return a \code{matrix} containing each pair of start and end points
//' on its columns, and the distance between these points in order to 
//' produce the Minimum Spanning Tree.
//'
// [[Rcpp::export]]
Rcpp::NumericMatrix mlpack_mst(arma::mat data)
{
	// transposing the matrix so mlpack accepts the format v4
	data = data.t();
	
	DualTreeBoruvka<> dtb(data); //initialization -- create the tree
	
	// find the MST
	
	arma::mat mstResults;
	dtb.ComputeMST(mstResults);
	// different than R logic -- it saves the results in mstResults
	
	// adding 1 to rows "from" & "to" - v3
	mstResults.row(0) = mstResults.row(0) + 1; //start index = 0 in C++
	mstResults.row(1) = mstResults.row(1) + 1;
	
	// returning a matrix instead of a list - v2
	return Rcpp::wrap(mstResults); 
}


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <RcppMLPACK.h>
#include <RcppArmadillo.h>
#include <Rcpp.h>

using namespace Rcpp;

// mlpack_mst
Rcpp::NumericMatrix mlpack_mst(arma::mat data);
RcppExport SEXP _emstreeR_mlpack_mst(SEXP dataSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat >::type data(dataSEXP);
    rcpp_result_gen = Rcpp::wrap(mlpack_mst(data));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_emstreeR_mlpack_mst", (DL_FUNC) &_emstreeR_mlpack_mst, 1},
    {NULL, NULL, 0}
};

RcppExport void R_init_emstreeR(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/stat-MST.R
#' @title Euclidean Minimum Spanning Tree Stat Function
#'
#' @description A Stat extension for 'ggplot2' to plot a 2D MST by making a 
#'     scatter plot with segments.
#' @description \code{stat_MST} uses the information returned by 
#'    \code{\link[emstreeR]{ComputeMST}} for producing a 2D Minimum Spanning 
#'    Tree plot with 'ggplot2' and should be combined with \code{geom_point()}.
#'    
#' @param mapping The aesthetic mapping, usually constructed with
#'   \code{\link[ggplot2]{aes}} or \code{\link[ggplot2]{aes_}}. 
#'    The required aesthetics are \code{x}, \code{y}, \code{from}, and \code{to}. 
#'    Those are columns of the \code{mst} object returned by 
#'    \code{\link[emstreeR]{ComputeMST}}.
#' @param data a \code{mst} class object returned by the 
#'    \code{\link[emstreeR]{ComputeMST}} function.
#' @param geom The geometric object to display the data. The default value is
#'    \code{"segment"} in order to produce the edges between the vertices.
#' @param position The position adjustment to use for overlapping points on this
#'    layer
#' @param show.legend logical. Should this layer be included in the legends?
#'    \code{NA}, the default, includes if any aesthetics are mapped. \code{FALSE}
#'    never includes, and \code{TRUE} always includes.
#' @param inherit.aes If \code{FALSE}, overrides the default aesthetics, rather
#'    than combining with them. This is most useful for helper functions that
#'    define both data and aesthetics and shouldn't inherit behaviour from the
#'    default plot specification, e.g. \code{\link[ggplot2]{borders}}.
#' @param linetype an integer or name: \code{0 = "blank"}, \code{1 = "solid"},
#'    \code{2 = "dashed"}, \code{3 = "dotted"}, \code{4 = "dotdash"}, 
#'    \code{5 = "longdash"}, \code{6 = "twodash"}. The default for \code{'MST'} 
#'    objects is "dotted".
#' @param ... other arguments passed on to \code{\link[ggplot2]{layer}}. This
#'    can include aesthetics whose values you want to set, not map. See
#'    \code{\link[ggplot2]{layer}} for more details.
#' @param na.rm	a logical value indicating whether \code{NA} values should be 
#' stripped before the computation proceeds.
#'
#' @section Computed variables: \describe{ 
#'    \item{x}{x coordinates of the MST start points}
#'    \item{y}{y coordinates of the MST start points}
#'    \item{xend}{x coordinates of the MST end points}
#'    \item{yend}{y coordinates of the MST end points}
#'    }
#'
#' @examples
#' 
#' ## 2D artificial data:
#' set.seed(1984)
#' n <- 15
#' c1 <- data.frame(x = rnorm(n,-0.2, sd=0.2), y = rnorm(n,-2,sd=0.2))
#' c2 <- data.frame(x = rnorm(n,-1.1, sd=0.15), y = rnorm(n,-2,sd=0.3)) 
#' d <- rbind(c1, c2)
#' d <- as.data.frame(d)
#' 
#' ## MST:
#' out <- ComputeMST(d)
#' 
#' #1) simple plot
#' library(ggplot2)
#' ggplot(data = out, 
#'     aes(x = x, y = y, 
#'     from=from, to=to))+ 
#'     geom_point()+
#'     stat_MST(colour="red", linetype = 2)
#'     
#' #2) curved edges
#' library(ggplot2)
#' ggplot(data = out, 
#'     aes(x = x, y = y, 
#'     from=from, to=to))+ 
#'     geom_point()+
#'     stat_MST(geom="curve", colour="red", linetype = 2)
#'
#' \dontrun{
#' ## plotting MST on maps:
#' library(ggmap)
#' 
#' #3) honeymoon cruise example
#' # define ports
#' df.port_locations <- data.frame(location = c("Civitavecchia, Italy",
#'                                              "Genova, Italy",
#'                                              "Marseille, France",
#'                                              "Barcelona, Spain",
#'                                              "Tunis, Tunisia",
#'                                              "Palermo, Italy"), 
#'                                 stringsAsFactors = FALSE)
#' 
#' # get latitude and longitude
#' geo.port_locations <- geocode(df.port_locations$location, source = "dsk")
#' 
#' # combine data
#' df.port_locations <- cbind(df.port_locations, geo.port_locations)
#' 
#' # MST
#' out <- ComputeMST(df.port_locations[,2:3])
#' plot(out) #just to check
#' 
#' # Plot
#' #' map <- c(left = -8, bottom = 32, right = 20, top = 47)
#' 
#' get_stamenmap(map, zoom = 5) %>% ggmap()+
#'   stat_MST(data = out, 
#'            aes(x = lon, y = lat, from=from, to=to), 
#'            colour="red", linetype = 2)+
#'   geom_point(data = out, aes(x = lon, y = lat), size=3)
#' 
#' 
#' #4) World Map travels:
#' library(ggplot2)
#' library(ggmaps)
#' 
#' country_coords_txt <- "
#'    1     3.00000  28.00000       Algeria
#'    2    54.00000  24.00000           UAE
#'    3   139.75309  35.68536         Japan
#'    4    45.00000  25.00000 'Saudi Arabia'
#'    5     9.00000  34.00000       Tunisia
#'    6     5.75000  52.50000   Netherlands
#'    7   103.80000   1.36667     Singapore
#'    8   124.10000  -8.36667         Korea
#'    9    -2.69531  54.75844            UK
#'    10    34.91155  39.05901        Turkey
#'    11  -113.64258  60.10867        Canada
#'    12    77.00000  20.00000         India
#'    13    25.00000  46.00000       Romania
#'    14   135.00000 -25.00000     Australia
#'    15    10.00000  62.00000        Norway"
#'    
#'    
#'  d <- read.delim(text = country_coords_txt, header = FALSE, 
#'  quote = "'", sep = "", col.names = c('id', 'lon', 'lat', 'name'))
#'    
#'  out <- ComputeMST(d[,2:3])
#'  
#'  country_shapes <- geom_polygon(aes(x = long, y = lat, group = group),
#'    data = map_data('world'), fill = "#CECECE", color = "#515151", 
#'    size = 0.15)
#'    
#'  ggplot()+ country_shapes+
#'    stat_MST(geomdata = out, aes(x = lon, y = lat, from=from, to=to), colour="red", linetype = 2)+
#'    geom_point(data = out, aes(x = lon, y = lat), size=2)
#'}
#'
#' @export
#' 
stat_MST <- function(mapping = NULL, data = NULL, geom = "segment", 
                     position = "identity", na.rm = FALSE, 
                     linetype="dotted", show.legend = NA, 
                     inherit.aes = TRUE, ...) {
  ggplot2::layer(
    stat = StatMST, 
    data = data, 
    mapping = mapping, 
    geom = geom, 
    position = position, 
    show.legend = show.legend, 
    inherit.aes = inherit.aes,
    params = list(na.rm = na.rm, linetype = linetype, ...)
  )
}

#' @title 'ggproto' objects
#'
#' @description Internal use only definitions of 'ggproto' objects needed for
#'   geoms and stats.
#'
#' @name emstreeR-ggproto
#' @format NULL
#' @usage NULL
#' @keywords internal
NULL
#> NULL

#' @rdname emstreeR-ggproto
#' @format NULL
#' @usage NULL
#' @export
StatMST <- ggplot2::ggproto("StatMST", ggplot2::Stat,
                       compute_group = function(data, scales, ...) {
                         ## Compute the line segment endpoints
                         
                         # change the last row's from/to columns
                         # repeat the row above
                         # this is done in order to work with geom="curve"
                         # avoiding the Error in calcCurveGrob(x, x$debug) : 
                         #.. end points must not be identical
                         data[nrow(data), c("from", "to")] <- 
                           c(
                             data[1, "from"], data[1, "to"]
                             )
                         
                         x = data[data$from, 1]
                         y = data[data$from, 2]
                         xend = data[data$to, 1]
                         yend = data[data$to, 2]
                         
                         data.frame(x=x, y=y, xend=xend, yend=yend)
                         
                       },
                       required_aes = c("x", "y", "from", "to")
)
